src/mapping/parser/mod.rs

```diff
@@ -194,128 +194,42 @@ fn query_arithmetic_from_pair(pair: Pair<Rule>) -> Result<Box<dyn query::Functio
-                other
-                    .next()
-                    .ok_or(TOKEN_ERR)?
-                    .into_inner()
-                    .next()
-                    .ok_or(TOKEN_ERR)?,
-            )?;
-            Ok(Box::new(ParseTimestampFn::new(&format, query, None)?))
-        }
-        Rule::upcase => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(UpcaseFn::new(query)))
-        }
-        Rule::downcase => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(DowncaseFn::new(query)))
-        }
-        Rule::uuid_v4 => Ok(Box::new(UuidV4Fn::new())),
-        Rule::sha1 => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(Sha1Fn::new(query)))
-        }
-        Rule::md5 => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(Md5Fn::new(query)))
-        }
-        Rule::now => {
-            let pair = pair.into_inner().next().ok_or(TOKEN_ERR)?;
-            let query = query_arithmetic_from_pair(pair)?;
-            Ok(Box::new(NowFn::new(query)))
-        }
-        _ => unreachable!("parser should not allow other query_function child rules here"),
+fn function_from_pairs(mut pairs: Pairs<Rule>) -> Result<Box<dyn query::Function>> {
+    dbg!(&pairs);
+
+    let function_name = pairs.next().ok_or(TOKEN_ERR)?;
+
+    let arguments = pairs
+        .filter_map(|pair| match pair.as_rule() {
+            Rule::function_argument => Some(pair.into_inner().next().expect("TODO")),
+            rule => unreachable!(
+                "parser must only support function_argument in function signature: {:?}",
+                rule
+            ),
+        })
+        .map(|pair| match pair.as_rule() {
+            Rule::query_arithmetic_boolean => (None, query_arithmetic_from_pair(pair)),
+            Rule::named_argument => {
+                let mut pairs = pair.into_inner();
+                let name = pairs.next().expect("TODO").as_span().as_str();
+                let pair = pairs.next().expect("TODO");
+
+                (Some(name), query_arithmetic_from_pair(pair))
+            }
+            rule => unreachable!(
+                "parser must only support query_arithmetic_boolean and named_argument arguments: {:?}",
+                rule
+            ),
+        })
+        .map(|(key, query_result)| match query_result {
+            Ok(query) => Ok((key, query)),
+            Err(err) => Err(err),
+        })
+        .collect::<Result<_>>()?;
+
+    match function_name.as_span().as_str() {
+        "downcase" => Ok(Box::new(DowncaseFn::new(arguments)?)),
```

Comment:
I'm pushing the responsibility of checking argument validity down to the function initialization code. That seems like the most logical place to do so?

This does not include validating the actual values passed in, as that can only happen at runtime due to the possibility of interpolating event values

Reply:
Checking argument validity could be done here iff the function types can tell the parser the "shape" of their arguments.

Reply:
With the latest update, argument validity is checked by the parser, for as much as possible.

That is to say, the function arity, any missing required argument, or any unknown provided keyword arguments are checked at "compile" time.

At runtime, the final value of the argument is checked against the list of accepted parameter values.